File: src/application/controllers/MainMenu.java
package application.controllers;

import java.io.File;
import java.io.IOException;
import java.net.URL;
import java.nio.file.Files;
import java.nio.file.StandardOpenOption;
import java.util.ResourceBundle;

import application.ControllerInitilizer;
import application.Data;
import application.ROM;
import application.excel.Excel;
import javafx.event.EventHandler;
import javafx.fxml.FXML;
import javafx.fxml.FXMLLoader;
import javafx.scene.Parent;
import javafx.scene.Scene;
import javafx.scene.control.Button;
import javafx.scene.image.Image;
import javafx.scene.image.ImageView;
import javafx.scene.input.KeyCode;
import javafx.scene.input.KeyCodeCombination;
import javafx.scene.input.KeyCombination;
import javafx.scene.input.KeyEvent;
import javafx.scene.input.MouseEvent;
import javafx.scene.layout.AnchorPane;
import javafx.stage.FileChooser;
import javafx.stage.FileChooser.ExtensionFilter;
import javafx.stage.Modality;
import javafx.stage.Stage;
import javafx.stage.Window;
import javafx.stage.WindowEvent;
import application.models.lists.Lists;
import application.models.lists.states.DataState;

public class MainMenu
{

	private Window	       mainWindow;
	private static Stage   primaryStage;
    boolean		           debug;

    @FXML
    void initialize()
    {
	this.disableButtons(true);
	debug = false;
	initializeImages();
	ROM.setProgramDirectory();
    }

    private void disableButtons(boolean disable)
    {
	save.setDisable(disable);
	character.setDisable(disable);
	apprentice.setDisable(disable);
	spell.setDisable(disable);
	equipment.setDisable(disable);
	shop.setDisable(disable);
	monster.setDisable(disable);
    }

    private void initializeImages()
    {

	open.setGraphic(new ImageView(new Image(this.getClass().getResourceAsStream("images/Open Icon2.png"))));
	save.setGraphic(new ImageView(new Image(this.getClass().getResourceAsStream("images/save icon.png"))));
	character.setGraphic(new ImageView(new Image(this.getClass().getResourceAsStream("images/Lemele.png"))));
	apprentice.setGraphic(new ImageView(new Image(this.getClass().getResourceAsStream("images/Lejes.png"))));
	spell.setGraphic(new ImageView(new Image(this.getClass().getResourceAsStream("images/Laser 3.png"))));
	equipment.setGraphic(new ImageView(new Image(this.getClass().getResourceAsStream("images/Weapon & Armor Shop Characters 2.png"))));
	shop.setGraphic(new ImageView(new Image(this.getClass().getResourceAsStream("images/Item Shop Character.png"))));
	ImageView i = (ImageView) monster.getGraphic();
	i.setImage(new Image(this.getClass().getResourceAsStream("images/Brain.png")));
	sceneryImage.setImage(new Image(this.getClass().getResourceAsStream("images/Scenery Icon.png")));
	sceneryImage.fitWidthProperty().bind(primaryStage.widthProperty());
	sceneryImage.fitHeightProperty().bind(primaryStage.heightProperty());
	primaryStage.getIcons().add(new Image(this.getClass().getResourceAsStream("images/Lux.png")));
    }

    @FXML
    void getFileToOpen(MouseEvent event) throws IOException
    {
	FileChooser fc = new FileChooser();
	fc.setTitle("Choose file to open");
	fc.getExtensionFilters().add(new ExtensionFilter("SNES Files", "*.smc"));
	fc.setInitialDirectory(ROM.getProgramDirectory());

	File rom = fc.showOpenDialog(null);

	if (rom != null)
	{
	    this.disableButtons(false);

	    ROM.setROMDirectory(rom.getParentFile());
		ROM.openDefaultROM();
		ROM.openROM(rom);
		System.out.println("ROM directory is: "+ROM.getROMDirectory());
	}
    }



    @FXML
    void writeDataToDisk()
    {
	    ROM.saveROM();
    }

    @FXML
    void characterEditor()
    {
	this.openDialog("fxmls/CharacterEditor.fxml", "Character Editor", true);
    }

    @FXML
    void apprenticeEditor()
    {
	this.openDialog("fxmls/ApprenticeEditor.fxml", "Apprentice Editor", false);
    }

    @FXML
    void spellEditor()
    {
	this.openDialog("fxmls/SpellEditor.fxml", "Spell Editor", false);
    }

    @FXML
    void equipmentEditor()
    {
	this.openDialog("fxmls/EquipmentEditor.fxml", "Equipment Editor", false);
    }

    @FXML
    void shopEditor()
    {
	this.openDialog("fxmls/ShopEditor.fxml", "Shop Editor", false);
    }

    @FXML
    void monsterEditor()
    {
	this.openDialog("fxmls/MonsterEditor.fxml", "Monster Editor", false);
    }

    public void openDialog(String filename, String title, boolean resizable)
    {
	try
	{
	    // Load the fxml file and create a new stage for the popup dialog.
	    FXMLLoader fxmlLoader = new FXMLLoader(this.getClass().getResource(filename));

	    Parent root = fxmlLoader.load();
	    final ControllerInitilizer controller = fxmlLoader.getController();

	    // Create the dialog Stage.
	    Stage editorStage = new Stage();
	    editorStage.initOwner(mainWindow);
	    editorStage.setTitle(title);
	    // editorStage.initStyle(StageStyle.);

	    editorStage.initModality(Modality.WINDOW_MODAL);
	    editorStage.setResizable(true);
	    Scene scene = new Scene(root);
	    editorStage.setScene(scene);

	    editorStage.setOnCloseRequest(new EventHandler<WindowEvent>()
	    {
		@Override
		public void handle(WindowEvent arg0)
		{
		    controller.saveData();
		    controller.saveState();

		}
	    });
	    editorStage.addEventHandler(KeyEvent.KEY_PRESSED, new EventHandler<KeyEvent>()
	    {

		@Override
		public void handle(KeyEvent event)
		{
		    final KeyCodeCombination kb = new KeyCodeCombination(KeyCode.S, KeyCombination.CONTROL_DOWN);
		    if (kb.match(event))
		    {
			System.out.println("data saved from main menu");
			controller.saveData();
			writeDataToDisk();
		    }
		}

	    });

	    // Show the dialog and wait until the user closes it
	    double height = primaryStage.getHeight();
	    double width = primaryStage.getWidth();
	    primaryStage.hide();
	    editorStage.showAndWait();
	    editorStage = null;
	    primaryStage.setHeight(height);
	    primaryStage.setWidth(width);
	    primaryStage.show();

	}
	catch (IOException e)
	{
	    e.printStackTrace();
	}
    }



    public void setMainWindow(Window mainWindow)
    {
	this.mainWindow = mainWindow;
    }


    public static Stage getPrimaryStage()
    {
	return primaryStage;
    }

    public static void setPrimaryStage(Stage primaryStage)
    {
	MainMenu.primaryStage = primaryStage;
    }

    @FXML
    private ResourceBundle resources;
    @FXML
    private ImageView	   sceneryImage;
    @FXML
    private URL		   location;
    @FXML
    private Button	   spell;
    @FXML
    private Button	   character;
    @FXML
    private Button	   apprentice;
    @FXML
    private Button	   shop;
    @FXML
    private Button	   save;
    @FXML
    private Button	   equipment;
    @FXML
    private AnchorPane	   mainMenuPane;
    @FXML
    private Button	   open;
    @FXML
    private Button	   monster;
}

File: src/application/controllers/MainMenu.kt
package application.controllers

import application.ControllerInitilizer
import application.ROM
import application.Settings
import javafx.event.EventHandler
import javafx.fxml.FXML
import javafx.fxml.FXMLLoader
import javafx.scene.Parent
import javafx.scene.Scene
import javafx.scene.control.Button
import javafx.scene.image.Image
import javafx.scene.image.ImageView
import javafx.scene.input.KeyCode
import javafx.scene.input.KeyCodeCombination
import javafx.scene.input.KeyCombination
import javafx.scene.input.KeyEvent
import javafx.stage.*
import javafx.stage.FileChooser.ExtensionFilter
import java.io.IOException

class MainMenu
{

	private lateinit var mainWindow: Window
	internal var debug: Boolean = false

	@FXML fun initialize()
	{
		this.disableButtons(true)
		initializeImages()
		ROM.setProgramDirectory()
		ROM.settings.readSettings()
	}

	private fun disableButtons(disable: Boolean)
	{
		save.isDisable = disable
		character.isDisable = disable
		apprentice.isDisable = disable
		spell.isDisable = disable
		equipment.isDisable = disable
		shop.isDisable = disable
		monster.isDisable = disable
		text.isDisable = disable
	}

	private fun initializeImages()
	{

		open.graphic = ImageView(Image(this.javaClass.getResourceAsStream("images/Open_Icon2.png")))
		save.graphic = ImageView(Image(this.javaClass.getResourceAsStream("images/save_icon.png")))
		character.graphic = ImageView(Image(this.javaClass.getResourceAsStream("images/Lemele.png")))
		apprentice.graphic = ImageView(Image(this.javaClass.getResourceAsStream("images/Lejes.png")))
		spell.graphic = ImageView(Image(this.javaClass.getResourceAsStream("images/Laser_3.png")))
		equipment.graphic = ImageView(Image(this.javaClass.getResourceAsStream("images/Weapon_and_Armor_Shop_Characters_2.png")))
		shop.graphic = ImageView(Image(this.javaClass.getResourceAsStream("images/Item_Shop_Character.png")))
		val i = monster.graphic as ImageView
		i.image = Image(this.javaClass.getResourceAsStream("images/Brain.png"))
		sceneryImage.image = Image(this.javaClass.getResourceAsStream("images/Scenery_Icon.png"))
		sceneryImage.fitWidthProperty().bind(primaryStage!!.widthProperty())
		sceneryImage.fitHeightProperty().bind(primaryStage!!.heightProperty())
		primaryStage!!.icons.add(Image(this.javaClass.getResourceAsStream("images/Lux.png")))
	}

	@FXML @Throws(IOException::class) fun getFileToOpen()
	{
		val fc = FileChooser()
		fc.title = "Choose file to open"
		fc.extensionFilters.add(ExtensionFilter("SNES Files", "*.smc"))
		fc.initialDirectory = ROM.programDirectory

		val rom = fc.showOpenDialog(null)

		if (rom != null)
		{
			disableButtons(false)

			ROM.romDirectory = rom.parentFile
			if(debug)
			{ROM.openDefaultROM()}
			ROM.openROM(rom)
			System.out.println("ROM directory is: ${ROM.romDirectory.canonicalPath}")
		}
	}


	@FXML fun writeDataToDisk() = ROM.saveROM()
	@FXML fun characterEditor() = openDialog("fxmls/CharacterEditor.fxml", "Character Editor", true)
	@FXML fun apprenticeEditor() = openDialog("fxmls/ApprenticeEditor.fxml", "Apprentice Editor", false)
	@FXML fun spellEditor() = openDialog("fxmls/SpellEditor.fxml", "Spell Editor", false)
	@FXML fun equipmentEditor() = openDialog("fxmls/EquipmentEditor.fxml", "Equipment Editor", false)
	@FXML fun shopEditor() = openDialog("fxmls/ShopEditor.fxml", "Shop Editor", false)
	@FXML fun monsterEditor() = openDialog("fxmls/MonsterEditor.fxml", "Monster Editor", false)
	@FXML fun textEditor() = openDialog("fxmls/TextEditor.fxml", "Text Editor", false)
	@FXML fun settingsEditor() = openDialog("fxmls/Settings.fxml", "Settings Editor", false)

	fun openDialog(filename: String, title: String, resizable: Boolean)
	{
		try
		{
			// Load the fxml file and create a new stage for the popup dialog.
			val fxmlLoader = FXMLLoader(this.javaClass.getResource(filename))

			val root = fxmlLoader.load<Parent>()
			val controller = fxmlLoader.getController<ControllerInitilizer>()

			// Create the dialog Stage.
			val editorStage: Stage? = Stage()
			editorStage!!.initOwner(mainWindow)
			editorStage.title = title
			// editorStage.initStyle(StageStyle.);

			editorStage.initModality(Modality.WINDOW_MODAL)
			editorStage.isResizable = true
			val scene = Scene(root)
			editorStage.scene = scene

			editorStage.onCloseRequest = EventHandler<WindowEvent> {
				controller.saveData()
				controller.saveState()
			}
			editorStage.addEventHandler(KeyEvent.KEY_PRESSED) { event ->
				val kb = KeyCodeCombination(KeyCode.S, KeyCombination.CONTROL_DOWN)
				if (kb.match(event))
				{
					println("data saved from main menu")
					controller.saveData()
					writeDataToDisk()
				}
			}

			// Show the dialog and wait until the user closes it
			val height = primaryStage!!.height
			val width = primaryStage!!.width
			primaryStage!!.hide()
			editorStage.showAndWait()
			primaryStage!!.height = height
			primaryStage!!.width = width
			primaryStage!!.show()

		}
		catch (e: IOException)
		{
			e.printStackTrace()
		}

	}

	fun setMainWindow(mainWindow: Window)
	{
		this.mainWindow = mainWindow
	}

	@FXML lateinit var sceneryImage: ImageView
	@FXML lateinit var spell: Button
	@FXML lateinit var character: Button
	@FXML lateinit var apprentice: Button
	@FXML lateinit var shop: Button
	@FXML lateinit var save: Button
	@FXML lateinit var equipment: Button
	@FXML lateinit var open: Button
	@FXML lateinit var monster: Button
	@FXML lateinit var text: Button
	@FXML lateinit var settings: Button

	companion object
	{
		var primaryStage: Stage? = null
	}
}